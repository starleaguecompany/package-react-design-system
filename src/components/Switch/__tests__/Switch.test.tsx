import * as React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Switch } from '../../../index';

describe('Switch', () => {
  test('should render correctly with no props', () => {
    render(<Switch />);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByTestId('Switch')).toHaveClass('container');
  });

  test('should render correctly with attributes', () => {
    render(<Switch id="test-id" className="test-class" />);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByTestId('Switch')).toHaveClass('test-class');
    expect(screen.getByRole('checkbox')).toHaveAttribute('id', 'test-id');
  });

  test('should render correctly with label', () => {
    render(<Switch>Test switch</Switch>);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByText('Test switch')).toBeVisible();
    expect(screen.getByText('Test switch')).toHaveClass('label');
  });

  test('should render correctly with checked prop', () => {
    const { container } = render(<Switch checked />);
    const checked = container.querySelectorAll('.checked');

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByTestId('Switch')).toHaveClass('checked');
    expect(checked.length).toEqual(3);
    expect(screen.getByRole('checkbox')).toBeChecked();
  });

  test('should render correctly with disabled prop', () => {
    const { container } = render(<Switch disabled />);
    const disabled = container.querySelectorAll('.disabled');

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByTestId('Switch')).toHaveClass('disabled');
    expect(disabled.length).toEqual(3);
    expect(screen.getByRole('checkbox')).toBeDisabled();
  });

  test('should render correctly with name prop', () => {
    render(<Switch name="Test switch" />);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByRole('checkbox')).toHaveAttribute('name', 'Test switch');
  });

  test('should render correctly with value prop', () => {
    render(<Switch value="5" />);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByRole('checkbox')).toHaveAttribute('value', '5');
  });

  test('should render correctly with readOnly prop', () => {
    render(<Switch readOnly />);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    expect(screen.getByTestId('Switch')).toHaveClass('readOnly');
  });

  test('should execute onChange callback on click switch', () => {
    const onChangeCallback = jest.fn();
    render(<Switch value="5" onChange={onChangeCallback} />);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    userEvent.click(screen.getByRole('checkbox'));
    expect(onChangeCallback).toBeCalledTimes(1);
  });

  test('should render correctly on external value update', () => {
    const onChangeCallback = jest.fn();
    const { rerender } = render(<Switch checked={false} onChange={onChangeCallback} />);

    expect(screen.getByTestId('Switch')).toMatchSnapshot();
    rerender(<Switch checked={true} onChange={onChangeCallback} />);
    expect(screen.getByRole('checkbox')).toBeChecked();
    expect(onChangeCallback).toBeCalledTimes(0);
  });
});
