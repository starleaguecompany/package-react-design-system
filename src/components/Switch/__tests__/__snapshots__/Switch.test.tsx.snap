// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should execute onChange callback on click switch 1`] = `
<label
  class="container"
  data-qa="Switch"
>
  <input
    class="input"
    type="checkbox"
    value="5"
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
</label>
`;

exports[`Switch should render correctly on external value update 1`] = `
<label
  class="container"
  data-qa="Switch"
>
  <input
    class="input"
    type="checkbox"
    value=""
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
</label>
`;

exports[`Switch should render correctly with attributes 1`] = `
<label
  class="test-class container"
  data-qa="Switch"
>
  <input
    class="input"
    id="test-id"
    type="checkbox"
    value=""
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
</label>
`;

exports[`Switch should render correctly with checked prop 1`] = `
<label
  class="container checked"
  data-qa="Switch"
>
  <input
    checked=""
    class="input"
    type="checkbox"
    value=""
  />
  <div
    class="control checked"
  >
    <div
      class="handler checked"
    />
  </div>
</label>
`;

exports[`Switch should render correctly with disabled prop 1`] = `
<label
  class="container disabled"
  data-qa="Switch"
>
  <input
    class="input"
    disabled=""
    type="checkbox"
    value=""
  />
  <div
    class="control disabled"
  >
    <div
      class="handler disabled"
    />
  </div>
</label>
`;

exports[`Switch should render correctly with label 1`] = `
<label
  class="container"
  data-qa="Switch"
>
  <input
    class="input"
    type="checkbox"
    value=""
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
  <span
    class="label"
  >
    Test switch
  </span>
</label>
`;

exports[`Switch should render correctly with name prop 1`] = `
<label
  class="container"
  data-qa="Switch"
>
  <input
    class="input"
    name="Test switch"
    type="checkbox"
    value=""
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
</label>
`;

exports[`Switch should render correctly with no props 1`] = `
<label
  class="container"
  data-qa="Switch"
>
  <input
    class="input"
    type="checkbox"
    value=""
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
</label>
`;

exports[`Switch should render correctly with readOnly prop 1`] = `
<label
  class="container readOnly"
  data-qa="Switch"
>
  <input
    class="input"
    type="checkbox"
    value=""
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
</label>
`;

exports[`Switch should render correctly with value prop 1`] = `
<label
  class="container"
  data-qa="Switch"
>
  <input
    class="input"
    type="checkbox"
    value="5"
  />
  <div
    class="control"
  >
    <div
      class="handler"
    />
  </div>
</label>
`;
